src/com/google/enterprise/adaptor/filenet/DocumentTraverser.java

```diff
@@ -317,11 +319,29 @@ private boolean hasPermissions(Permissions.Acl permissions,
   // security template in particular, so that SecurityPolicyTraverser can
   // update just the one ACL and not have to rewire the chain?
   private void processInheritedPermissions(DocId docId,
-      Permissions.Acl permissions, Response response) {
+      ActiveMarkingList activeMarkings, Permissions.Acl permissions,
+      Response response) {
     // Send add request for adding ACLs inherited from parent folders.
     String secParentFragment = null;
-    Acl folderAcl =
-        createAcl(docId, permissions, PermissionSource.SOURCE_PARENT, null);
+    Iterator<?> iterator = (activeMarkings == null)
+        ? Collections.emptyIterator() : activeMarkings.iterator();
+    while (iterator.hasNext()) {
+      Marking marking = ((ActiveMarking) iterator.next()).get_Marking();
+      Permissions.Acl markingPerms =
+          new Permissions(marking.get_Permissions()).getAcl();
+      Acl markingAcl = createAcl(docId, secParentFragment,
+          Acl.InheritanceType.AND_BOTH_PERMIT, options.getGlobalNamespace(),
+          markingPerms.getAllowUsers(), markingPerms.getDenyUsers(),
+          markingPerms.getAllowGroups(), markingPerms.getDenyGroups());
+      if (markingAcl != null) {
```

Comment:
Why not log absent active markings here? Maybe re-unify these code blocks in the followup CL (setting secParentFragment first everywhere, etc.)?